### PR TITLE
Userns deleg translate ids

### PR DESCRIPTION
**Pull request by Blub**
Translate userids in `zfs allow` ioctrls, so that within a user namespace one can reference namespace-local user IDs.
In order for better testing this also adds support for specifying numeric IDs.

### Motivation and Context
Without this unprivileged containers an inspect the ZFS datasets and even use `zfs allow`, but they have to manually map their user ids. And since no numeric IDs can be used they'd need to temporarily modify /etc/passwd to reflect the user namespace mapping.

### Description
-) Add support for numeric IDs to zfs allow/unallow
-) Add helpers to perform UID translation on nvlists
-) Make `zfs_ioc_get_fsacl()` use those helpers
-) Allow a process with `CAP_SYS_ADMIN` to delegate privileges to userids that are mapped in the current user namespace.

### How Has This Been Tested?
Tested various operations in unprivileged lxc containers, and added some test cases.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
^ Arguable...
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [?] All new and existing tests passed.
(getting unexpected failures with unmodified git-master already...)
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by Blub** [comment]
Updating with the 80char fix and hopefully buildfixes for the centos version. Not exactly sure about that 16.04 one.

File: module/zfs/zfs_ioctl.c [hunk 1]
```diff
@@ -3016,6 +3155,56 @@ zfs_ioc_set_fsacl(zfs_cmd_t *zc)
...
 }
 
+#if defined(CONFIG_USER_NS) && defined(HAVE_KUIDGID_T)
+/*
+ * On success this consumes the nvlist in pnvp and replaces the pointer with
+ * one where ids have been mapped into the current user namespace.
+ * In case of error, pnvpn and its pointee stay unmodified.
+ */
+static int
+deleg_map_user_ns(nvlist_t **pnvp)
```
Comment (behlendorf) [inline module/zfs/zfs_ioctl.c]: This should probably be renamed `dsl_deleg_map_user_ns()` and moved it in to `dsl_deleg.c` along with the `fsacl_map_*` functions.  The code added to `zfs_ioc_set_fsacl` should probably also be there there as well in it's own function.

Let's also make sure that these functions exist and `return 0` when `CONFIG_USER_NS && HAVE_KUIDGID_T` aren't defined.  This should help keep this a little more readable and portable.

**Comment by Blub** [comment]
Thanks for the feedback! Will try to find some time for the changes soon.

**Comment by ahrens** [comment]
I noticed that this PR hasn’t been updated in some time.  We would like to see this feature added, but it seems that it isn’t quite complete, so we’re going to close it for now.  When you have time to revisit this work, feel free to reopen this PR or open a new one.  Thanks for your contribution to OpenZFS and please continue to open PR’s to address issues that you encounter!

**Pull request closed without merging**